File: answers/leetcode/Evaluate Reverse Polish Notation/Evaluate Reverse Polish Notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            int number0;
            int number1;
            switch (token) {
                case "+":
                    number0 = stack.pop();
                    number1 = stack.pop();
                    stack.push(number0 + number1);
                    break;
                case "-":
                    number0 = stack.pop();
                    number1 = stack.pop();
                    stack.push(number1 - number0);
                    break;
                case "*":
                    number0 = stack.pop();
                    number1 = stack.pop();
                    stack.push(number0 * number1);
                    break;
                case "/":
                    number0 = stack.pop();
                    number1 = stack.pop();
                    stack.push(number1 / number0);
                    break;
                default:
                    stack.push(Integer.parseInt(token));
                    break;
            }
        }
        return stack.isEmpty() ? 0 : stack.pop();
    }
}
//Runtime: 4 ms, faster than 93.38% of Java online submissions for Evaluate Reverse Polish Notation.
//Memory Usage: 38.8 MB, less than 59.87% of Java online submissions for Evaluate Reverse Polish Notation.

File: answers/leetcode/Evaluate Reverse Polish Notation/Evaluate Reverse Polish Notation.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """

        num_stack = []
        for i in range(len(tokens)):
            if '+' == tokens[i] or '-' == tokens[i] or '*' == tokens[i] or '/' == tokens[i]:
                num0 = num_stack.pop()
                num1 = num_stack.pop()
                num_stack.append(self.calculate(num1, num0, tokens[i]))
            else:
                num_stack.append(int(tokens[i]))
        return num_stack[-1]

    def calculate(self, num0, num1, op):
        answer = 0
        if '+' == op:
            answer = num0 + num1
        elif '-' == op:
            answer = num0 - num1
        elif '*' == op:
            answer = num0 * num1
        elif '/' == op:
            answer = abs(num0) / abs(num1)
            if num0 * num1 < 0:
                answer *= -1
        return answer
